packages/framework/src/Model/Product/ProductChangeMessageProducer.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\FrameworkBundle\Model\Product;
+
+use OldSound\RabbitMqBundle\RabbitMq\ProducerInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+
+class ProductChangeMessageProducer
+{
+    /**
+     * @var int[]
+     */
+    protected $scheduledProductIds = [];
+
+    /**
+     * @var \OldSound\RabbitMqBundle\RabbitMq\ProducerInterface
+     */
+    protected $productChangeProducer;
+
+    /**
+     * @param \OldSound\RabbitMqBundle\RabbitMq\ProducerInterface $productChangeProducer
+     */
+    public function __construct(ProducerInterface $productChangeProducer)
+    {
+        $this->productChangeProducer = $productChangeProducer;
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     */
+    public function productChanged(Product $product): void
```

Comment:
I think that updating all associated products is good way to avoid inconsistent data.

You are right that behavior should be same so I extended productChangedByIds method

Reply:
d401af4902a30793b81818334db7f751fa5dcac3

Reply:
and fixed in e1d0eeb0d79988c8900e5dc85b62438d399d5d35

Reply:
Now, when you call `productChanged()` with a `Product` with ID 1 and variants with IDs 2, 3, and 4, it will call `productChangedByIds()` with `[1,2,3,4]` and that will fetch the four products one by one from the repository, create the array `[1,2,3,4]` four times and merge it together and deduplicate.

It seems unnecessarily complex.

The logic should be in one of those 4 public methods and the other 3 should just transform the parameter and delegate.

Furthermore, this class seems like a good candidate for unit testing. The logic of associated products seems like a good candidate for class extraction. It's basically copy-pasted in `ProductSellingDeniedRecalculator::getProductsForCalculations()` and `Product::markForVisibilityRecalculation()` - maybe create an issue for that?